src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/WebSocketStateHelper.cs

```diff
@@ -0,0 +1,50 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Net.WebSockets
+{
+    internal static class WebSocketStateHelper
+    {
+        /// <summary>Valid states to be in when calling SendAsync.</summary>
+        internal const ManagedWebSocketStates ValidSendStates = ManagedWebSocketStates.Open | ManagedWebSocketStates.CloseReceived;
+        /// <summary>Valid states to be in when calling ReceiveAsync.</summary>
+        internal const ManagedWebSocketStates ValidReceiveStates = ManagedWebSocketStates.Open | ManagedWebSocketStates.CloseSent;
+        /// <summary>Valid states to be in when calling CloseOutputAsync.</summary>
+        internal const ManagedWebSocketStates ValidCloseOutputStates = ManagedWebSocketStates.Open | ManagedWebSocketStates.CloseReceived;
+        /// <summary>Valid states to be in when calling CloseAsync.</summary>
+        internal const ManagedWebSocketStates ValidCloseStates = ManagedWebSocketStates.Open | ManagedWebSocketStates.CloseReceived | ManagedWebSocketStates.CloseSent;
+
+        internal static bool IsValidSendState(WebSocketState state) => ValidSendStates.HasFlag(ToFlag(state));
+
+        internal static void ThrowIfInvalidState(WebSocketState currentState, bool isDisposed, Exception? innerException, ManagedWebSocketStates validStates)
+        {
+            ManagedWebSocketStates state = ToFlag(currentState);
+
+            if (!validStates.HasFlag(state))
```

Comment:
We generally prefer to avoid using HasFlag, even though the JIT does optimize away the boxing in tier 1:
```suggestion
            if ((state & validStates) == 0)
```